MatrixSDK/Space/MXSpaceNotificationCounter.swift

```diff
@@ -0,0 +1,160 @@
+        sum.directMissedDiscussionsCount = left.directMissedDiscussionsCount + right.directMissedDiscussionsCount
+        sum.directMissedDiscussionsHighlightedCount = left.directMissedDiscussionsHighlightedCount + right.directMissedDiscussionsHighlightedCount
+        sum.groupMissedDiscussionsCount = left.groupMissedDiscussionsCount + right.groupMissedDiscussionsCount
+        sum.groupMissedDiscussionsHighlightedCount = left.groupMissedDiscussionsHighlightedCount + right.groupMissedDiscussionsHighlightedCount
+        return sum
+    }
+    
+    static public func +=( left: inout MXSpaceNotificationState, right: MXSpaceNotificationState) {
+        left = left + right
+    }
+}
+
+@objcMembers
+public class MXSpaceNotificationCounter: NSObject {
+    
+    // MARK: - Properties
+    
+    public var homeNotificationState = MXSpaceNotificationState()
+    private var notificationStatePerSpaceId: [String:MXSpaceNotificationState] = [:]
+    
+    // MARK: - Public
+    
+    /// Compute the notification count for every spaces
+    /// - Parameters:
+    ///   - spaces: list of spaces of the current sessiom
+    ///   - rooms: list of rooms of the current session
+    ///   - flattenedParentIds: the flattened aprent ID for all rooms
+    public func computeNotificationCount(for spaces:[MXSpace], with rooms:[MXRoom], flattenedParentIds: [String: Set<String>]) {
+        let startDate = Date()
+        MXLog.debug("[Spaces] computeNotificationCount started")
+
+        var homeNotificationState = MXSpaceNotificationState()
+        var notificationStatePerSpaceId: [String:MXSpaceNotificationState] = [:]
+        
+        for room in rooms {
+            guard room.summary.roomType != .space else {
+                continue
+            }
+            
+            let notificationState = self.notificationState(forRoomWithId: room)
+            
+            homeNotificationState += notificationState
+            for space in spaces {
+                if flattenedParentIds[room.roomId]?.contains(space.spaceId) ?? false {
+                    let storedState = notificationStatePerSpaceId[space.spaceId] ?? MXSpaceNotificationState()
+                    notificationStatePerSpaceId[space.spaceId] = storedState + notificationState
+                }
+            }
+        }
+        
+        self.homeNotificationState = homeNotificationState
+        self.notificationStatePerSpaceId = notificationStatePerSpaceId
+        
+        MXLog.debug("[Spaces] computeNotificationCount ended after \(Date().timeIntervalSince(startDate))s")
+    }
+    
+    /// Notification state for a given space
+    /// - Parameters:
+    ///   - spaceId: ID of the space
+    /// - Returns: a `MXSpaceNotificationState` instance with the number of notifications for the given space
+    public func notificationState(forSpaceWithId spaceId: String) -> MXSpaceNotificationState {
+        return notificationStatePerSpaceId[spaceId] ?? MXSpaceNotificationState()
```

Comment:
Should we create in this case the instance in the dictionary? Because we are returning an instance that will not be the same the next time it will be in `notificationStatePerSpaceId ` dict.